Regression(254490@main): null dereference in Permissions::query

szewai opened the pull request:
#### 7dc6d6b06a8c4df7f633f39a496b162035dc17ab
<pre>
Regression(254490@main): null dereference in Permissions::query
<a href="https://bugs.webkit.org/show_bug.cgi?id=247304">https://bugs.webkit.org/show_bug.cgi?id=247304</a>
rdar://101789656

Reviewed by Wenson Hsieh and Alex Christensen.

When creating PermissionStatus in Permissions::query(), we did not null check document-&gt;page() before dereferencing it.

* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::query):

Canonical link: <a href="https://commits.webkit.org/256209@main">https://commits.webkit.org/256209@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a69f3b06f4984fbb1967063fbdf6aec6663f7d30

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95068 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4211 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27985 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104670 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164922 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99058 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4302 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/32398 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87377 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100570 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100735 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3130 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81598 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30111 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/86877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73006 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38801 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36623 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/19713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4293 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40557 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42536 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38954 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -148,23 +148,29 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
 
     if (document) {
-        if (!document->page()) {
+        WeakPtr<Page> page = document->page();
```
```suggestion
        WeakPtr page = document->page();
```

rniwa commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -148,23 +148,29 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
             }
 
-            promise.resolve(PermissionStatus::create(document, *permissionState, permissionDescriptor, PermissionQuerySource::Window, *document->page()));
+            if (!page || page.get() != document->page()) {
```
I don't think we need to check page != document->page()
A `Document` cannot move from one `Page` to another.

szewai replied:
Sure, will remove

szewai committed:
Regression(254490@main): null dereference in Permissions::query
https://bugs.webkit.org/show_bug.cgi?id=247304
rdar://101789656

Reviewed by Wenson Hsieh and Alex Christensen.

When creating PermissionStatus in Permissions::query(), we did not null check document->page() before dereferencing it.

* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::query):

Canonical link: https://commits.webkit.org/256209@main
